test_conformance/basic/test_vector_swizzle.cpp

```diff
@@ -0,0 +1,680 @@
+    test_error(error, "Unable to enqueue test kernel");
+
+    error = clFinish(queue);
+    test_error(error, "clFinish failed after test kernel");
+
+    error =
+        clEnqueueReadBuffer(queue, mem, CL_TRUE, 0, buffer.size() * sizeof(T),
+                            buffer.data(), 0, NULL, NULL);
+    test_error(error, "Unable to read data after test kernel");
+
+    if (buffer != reference)
+    {
+        log_error("Result buffer did not match reference buffer!\n");
+        return TEST_FAIL;
+    }
+
+    return TEST_PASS;
+}
+
+template <typename T, size_t N>
+static int test_vectype(const char* type_name, cl_device_id device,
+                        cl_context context, cl_command_queue queue)
+{
+    log_info("    testing type %s%d\n", type_name, N);
+
+    cl_int error = CL_SUCCESS;
+    int result = TEST_PASS;
+
+    clProgramWrapper program;
+    clKernelWrapper kernel;
+
+    std::string buildOptions{ "-DTYPE=" };
+    buildOptions += type_name;
+    buildOptions += std::to_string(N);
+    buildOptions += " -DBASETYPE=";
+    buildOptions += type_name;
+
+    std::vector<T> value(N);
+    std::iota(value.begin(), value.end(), 0);
+
+    std::vector<T> reference;
+    constexpr size_t S = TestInfo<N>::vector_size;
+    makeReference<T, N, S>(reference);
+
+    // XYZW swizzles:
+
+    const char* xyzw_source = TestInfo<N>::kernel_source_xyzw;
+    error = create_single_kernel_helper(
+        context, &program, &kernel, 1, &xyzw_source, "test_vector_swizzle_xyzw",
+        buildOptions.c_str());
+    test_error(error, "Unable to create xyzw test kernel");
+
+    result |= test_vectype_case(value, reference, context, kernel, queue);
+
+    // sN swizzles:
+    const char* sN_source = TestInfo<N>::kernel_source_sN;
+    error = create_single_kernel_helper(context, &program, &kernel, 1,
+                                        &sN_source, "test_vector_swizzle_sN",
+                                        buildOptions.c_str());
+    test_error(error, "Unable to create sN test kernel");
+
+    result |= test_vectype_case(value, reference, context, kernel, queue);
+
+    // RGBA swizzles for OpenCL 3.0 and newer:
+    const Version device_version = get_device_cl_version(device);
+    if (device_version >= Version(3, 0))
+    {
+        const char* rgba_source = TestInfo<N>::kernel_source_rgba;
+        error = create_single_kernel_helper(
+            context, &program, &kernel, 1, &rgba_source,
+            "test_vector_swizzle_rgba", buildOptions.c_str());
+        test_error(error, "Unable to create rgba test kernel");
+
+        result |= test_vectype_case(value, reference, context, kernel, queue);
+    }
+
+    return result;
+}
+
+template <typename T>
+static int test_type(const char* type_name, cl_device_id device,
+                     cl_context context, cl_command_queue queue)
+{
+    return test_vectype<T, 2>(type_name, device, context, queue)
+        | test_vectype<T, 3>(type_name, device, context, queue)
```

Comment:
Qualcomm has concerns with the SetKernelArg usage in this PR, specifically for 3-component vectors. 

As per OpenCL-C spec (page 11) "For 3-component vector data types, the size of the data type is 4 * sizeof(component)". Furthermore, from the OpenCL-API spec for clSetKernelArg error conditions, "CL_INVALID_ARG_SIZE if arg_size does not match the size of the data type for an argument that is not a memory object"

This would imply that for 3-component vectors, the size passed to SetKernelArg should be 4 * sizeof(component).

Reply:
Ah, good catch.  I'm not sure why this wasn't failing on our implementation.

I just pushed a commit that uses the special `vector_size` instead of the number of components in the vector.  Can you please check that this fixes this issue?

Reply:
Yes, the vector_swizzle test now passes with your patch. Thanks for fixing.